DataDog/dd-trace-rb
# OpenTelemetry Traces support

**What does this PR do?**

_The PR is divided into compartmentalized commits, if you prefer reviewing one commit at a time._

This PR allows `ddtrace` users to use Datadog Tracing and [OpenTelemetry Traces](https://opentelemetry.io/docs/concepts/signals/traces/) together.

OpenTelemetry Traces created by the host application will be collected by `ddtrace` and sent to the Datadog UI without any code changes.

Datadog and OpenTelemetry spans will form a single trace, with correct parenting being handled by `ddtrace`.

To activate the OpenTelemetry Traces collection, run this line:

```ruby
require 'datadog/opentelemetry'
```

**Motivation**


OpenTelemetry is a vendor-agnostic instrumentation specification that is being widely supported across the tech community today.

We should support users that want to keep using this vendor-agnostic API while using Datadog to process the telemetry data created.

**Additional Notes**

**This is the first version of OpenTelemetry Traces support. It will be released as experimental.**

This PR includes:
* Testing: Unit test coverage.
* Testing: An integration app exercising OpenTelemetry trace interweaving.
* Testing: Run the OpenTelemetry Ruby repository test suite with `ddtrace` integration enabled (`bundle exec rake upstream:opentelemetry`).
* Documentation about how to configure Datadog APM with OpenTelemetry, including current limitations.
* Telemetry collection for OpenTelemetry (a boolean flag for when OpenTelemetry integration is enabled).

**Examples (from the `integration` app)**

OpenTelemetry root span, with a Datadog child span:

<img width="761" alt="Screen Shot 2023-01-13 at 10 30 11 AM" src="https://user-images.githubusercontent.com/583503/212393345-bd039a4f-e65f-4e74-b680-23af78034536.png">

Datadog root span, with an OpenTelemetry child span:

<img width="759" alt="Screen Shot 2023-01-13 at 10 29 47 AM" src="https://user-images.githubusercontent.com/583503/212393371-6e25a0d3-3ec0-463d-96fd-d2b77f5a9310.png">

And because we handle trace parenting now, a baseline test case to ensure this PR didn't break pure OpenTelemetry parenting:

<img width="754" alt="Screen Shot 2023-01-13 at 10 30 38 AM" src="https://user-images.githubusercontent.com/583503/212393424-c1aeff21-9cea-4818-a708-fe09181f434f.png">


**How to test the change?**

* The easiest way is to run the `integration/apps/opentelemetry` integration app, to see real spans in the Datadog UI.
* `bundle exec rake upstream:opentelemetry` ensures no OpenTelemetry contract has been violated by this integration.
